sycamore/llms/openai.py

```diff
@@ -0,0 +1,252 @@
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        azure_deployment: Optional[str] = None,
+        azure_endpoint: Optional[str] = None,
+        api_version: Optional[str] = None,
+        azure_ad_token: Optional[str] = None,
+        azure_ad_token_provider: Optional[AzureADTokenProvider] = None,
+        # Deprecated names that we support for backwards compatibility.
+        api_type: Optional[str] = None,
+        api_base: Optional[str] = None,
+        # Additional OpenAI Client parameters that will be passed in.
+        **kwargs,
+    ):
+        if api_type is not None:
+            logger.warn("WARNING: The api_type parameter is deprecated. Please use client_type instead.")
+            if api_type in {"azure", "azure_ad", "azuread"}:
+                client_type = OpenAIClientType.AZURE
+            else:
+                client_type = OpenAIClientType.OPENAI
+
+        if api_base is not None:
+            logger.warn("WARNING: The api_base parameter is deprecated. Please use base_url or azure_endpoint instead.")
+
+            if azure_endpoint is None:
+                azure_endpoint = api_base
+            else:
+                raise ValueError("Can't set both api_base and azure_endpoint")
+
+        # TODO: Add some parameter validation so we can fail fast. The openai library has a bunch of validation,
+        # but may not happen until much later in the job execution.
+
+        self.client_type = client_type
+        self.api_key = api_key
+        self.base_url = base_url
+        self.organization = organization
+        self.max_retries = max_retries
+        self.azure_deployment = azure_deployment
+        self.azure_endpoint = azure_endpoint
+        self.api_version = api_version
+        self.azure_ad_token = azure_ad_token
+        self.azure_ad_token_provider = azure_ad_token_provider
+        self.extra_kwargs = kwargs
+
+        # The OpenAI Python library is happy to pull Azure creds from the AZURE_OPENAI_API_KEY environment variable,
+        # but Guidance will error out if neither api_key nor azure_ad_token_provider are explicitly set.
+        if client_type == OpenAIClientType.AZURE and api_key is None and azure_ad_token_provider is None:
+            if "AZURE_OPENAI_API_KEY" not in os.environ:
+                raise ValueError(
+                    "Must set either api_key, azure_ad_token_provider, or AZURE_OPENAI_API_KEY environment variable."
+                )
+
+            self.api_key = os.environ.get("AZURE_OPENAI_API_KEY")
+
+    def get_client(self) -> OpenAIClient:
+        if self.client_type == OpenAIClientType.OPENAI:
+            return OpenAIClient(
+                api_key=self.api_key,
+                organization=self.organization,
+                base_url=self.base_url,
+                max_retries=self.max_retries,
+                **self.extra_kwargs,
+            )
+        elif self.client_type == OpenAIClientType.AZURE:
+            if self.azure_endpoint is None:
```

Comment:
shouldn't this just be in the constructor with other validations?

Reply:
yeah, let me change that. Not quite sure why I didn't put it there to start.